frontend/lib/src/components/elements/ArrowVegaLiteChart/useVegaEmbed.ts

```diff
@@ -0,0 +1,285 @@
+  finalizeView: () => void
+}
+
+export function useVegaEmbed(
+  inputElement: VegaLiteChartElement,
+  widgetMgr: WidgetStateManager,
+  fragmentId?: string
+): UseVegaEmbedOutput {
+  const vegaView = useRef<VegaView | null>(null)
+  const vegaFinalizer = useRef<(() => void) | null>(null)
+  const defaultDataName = useRef<string>(DEFAULT_DATA_NAME)
+  const dataRef = useRef<Quiver | null>(null)
+  const datasetsRef = useRef<WrappedNamedDataset[]>([])
+
+  // Setup interactivity for the chart if it supports selections
+  const { maybeConfigureSelections, onFormCleared } = useVegaLiteSelections(
+    inputElement,
+    widgetMgr,
+    fragmentId
+  )
+
+  useFormClearHelper({ widgetMgr, element: inputElement, onFormCleared })
+
+  const { id: chartId, data, datasets } = inputElement
+
+  // Initialize the data and datasets refs with the current data and datasets
+  // This is predominantly used to handle the case where we want to reference
+  // these in createView before the first render.
+  useEffect(() => {
+    if (vegaView.current === null) {
+      dataRef.current = data
+      datasetsRef.current = datasets
+    }
+  }, [data, datasets])
+
+  const finalizeView = useCallback(() => {
+    if (vegaFinalizer.current) {
+      vegaFinalizer.current()
+    }
+
+    vegaFinalizer.current = null
+    vegaView.current = null
+  }, [])
+
+  const createView = useCallback(
+    async (
+      containerRef: RefObject<HTMLDivElement>,
+      spec: any
+    ): Promise<VegaView | null> => {
+      logMessage("Creating a new Vega view.")
+
+      if (containerRef.current === null) {
+        throw new Error("Element missing.")
+      }
+
+      // Finalize the previous view so it can be garbage collected.
+      finalizeView()
+
+      const options = {
+        // Adds interpreter support for Vega expressions that is compliant with CSP
+        ast: true,
+        expr: expressionInterpreter,
+
+        // Disable default styles so that vega doesn't inject <style> tags in the
+        // DOM. We set these styles manually for finer control over them and to
+        // avoid inlining styles.
+        tooltip: { disableDefaultStyle: true },
+        defaultStyle: false,
+        forceActionsMenu: true,
+      }
+
+      const { vgSpec, view, finalize } = await embed(
+        containerRef.current,
+        spec,
+        options
+      )
+
+      vegaView.current = view
+
+      maybeConfigureSelections(view)
+
+      vegaFinalizer.current = finalize
+
+      // Try to load the previous state of the chart from the element state.
+      // This is useful to restore the selection state when the component is re-mounted
+      // or when its put into fullscreen mode.
+      const viewState = widgetMgr.getElementState(chartId, "viewState")
+      if (notNullOrUndefined(viewState)) {
+        try {
+          vegaView.current = view.setState(viewState)
+        } catch (e) {
+          logWarning("Failed to restore view state", e)
+        }
+      }
```

Comment:
I believe for `ArrowVegaLiteChart`, reconstructing this state is only relevant if selections are actually enabled/used (it's different for Plotly chart). Maybe this logic can just be moved into `maybeConfigureSelections`?